fs/copy_linux.go

```diff
@@ -55,7 +56,13 @@ func copyFileContent(dst, src *os.File) error {
 	dstFd := int(dst.Fd())
 
 	for size > 0 {
-		n, err := unix.CopyFileRange(srcFd, nil, dstFd, nil, int(size), 0)
+		var copySize int
+		if size > int64(math.MaxInt32) {
```

Comment:
Is this a limitation on all platforms?

Reply:
According to @cpuguy83's comment in https://github.com/containerd/continuity/commit/afba265aa60a6dab3bf3a7809b2e6e9a95704b04 it never copies more than 2GB even when it doesn't fail. Actually looking at the kernel source it looks like it is actually up to the fs implementation. That said I don't think there's going to exist any storage in the near future where one syscall per 2GB data is going to be a non-negligible overhead.

Now that I was looking at the kernel it actually seems like sizes larger than 2GB won't work on any 32-bit system, see https://github.com/torvalds/linux/blob/30b06abfb92bfd5f9b63ea6a2ffb0bd905d1a6da/fs/read_write.c#L372, in which case the commit message here is misleading (is the travis build only testing on amd64? That might be a problem).

Reply:
> That said I don't think there's going to exist any storage in the near future where one syscall per 2GB data is going to be a non-negligible overhead.

We added support for this call to make use of shared extents on BTRFS and XFS which can perform these copies very fast since there is no byte copying, hence my wondering of why this limit would have any relevance on such a system. I guess what I want to understand is what is `int(size)` doing on these 32 bit systems when a large value is given? Perhaps we need a smarter truncation function than just casting and if MIPS has specific logic (such as 64 bit but limits 32 bits with this syscall) we can have a MIPS specific implementation which does a 64->32 bit truncation.

Reply:
Well int(size) is truncating when int is 32-bit which causes it to just wraparound. But the documentation for copy_file_range claims that it takes a size_t which should be uint in go, but it would work anyways because it just gets reinterpreted as unsigned in the kernel. But then again in the kernel it is actually checked that it doesn't overflow an ssize_t (which doesn't seems to be explicitly documented, but the return value is ssize_t), so there is no way to exceed 2GB at a time on 32-bit systems.

What happens with respect to shared extents larger than 2GB I do not know, but hopefully the filesystems are smart enough to detect this. Or it might not even be an issue because it might actually be implemented as a number of smaller extents right after each other. Anyways the limitation on the syscall is what it is.

Just to verify I tried building on x86 (i.e. 32-bit) and the test also fails with EINVAL. So at least we need to ensure that the size given never exceeds SSIZE_MAX.

Reply:
Could we change the logic then to detect the truncation then

```
copySize := int(size)
if size != int64(copySize) && size > int64(math.MaxInt32) {
	copySize = int(math.MaxInt32)
}
```

What I don't understand is how can `int(size)` be truncating yet `int(size) > int(math.MaxInt32)`. I assume `int(size) < 0` cannot be true even with wraparound.

Reply:
int(size) > int(math.MaxInt32) can only be true on 64-bit archs where int is 64-bits, but anyways I changed it to explicitly check against the max value of int which corresponds to SSIZE_MAX.